bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -41,27 +41,36 @@ export type L3Mapping = Map<string, {value: number, mapping: L2Mapping}>
 
 export type Mapping = L1Mapping | L2Mapping | L3Mapping
 
-export function map_one_level(factors: L1Factor[], padding: number, offset: number = 0): [L1Mapping, number] {
+export type MappingSpec = {mapping: Mapping, tops: L1Factor[] | null, mids: L2Factor[] | null, inner_padding: number}
+
+export type MappingEntry = {value: number, mapping?: L1Mapping | L2Mapping}
+
+export function map_one_level(
+    factors: L1Factor[],
+    padding: number,
+    offset: number = 0): MappingSpec {
   const mapping: L1Mapping = new Map()
 
   for (let i = 0; i < factors.length; i++) {
     const factor = factors[i]
-    if (!mapping.has(factor)) {
-      mapping.set(factor, {value: 0.5 + i*(1 + padding) + offset})
-    } else {
+    if (mapping.has(factor)) {
       throw new Error(`duplicate factor or subfactor: ${factor}`)
     }
+    mapping.set(factor, {value: 0.5 + i*(1 + padding) + offset})
   }
 
-  return [mapping, (factors.length - 1)*padding]
+  const inner_padding = (factors.length - 1)*padding
+  return {mapping, tops: null, mids: null, inner_padding}
 }
 
-export function map_two_levels(factors: L2Factor[],
-                               outer_pad: number, factor_pad: number,
-                               offset: number = 0): [L2Mapping, number] {
+export function map_two_levels(
+    factors: L2Factor[],
+    outer_pad: number,
+    factor_pad: number,
+    offset: number = 0): MappingSpec {
```

Comment:
The common pattern I've been using across bokehjs for functions with multi-line argument lists, is like this:
```suggestion
export function map_two_levels(
  factors: L2Factor[],
  outer_pad: number,
  factor_pad: number,
  offset: number = 0,
): MappingSpec {
```